### DISCUSSION TITLE
Reconsider translation sets

### DISCUSSION THREAD

**Discussion by ocean90** [GlotPress/GlotPress-WP#1331] [General]
This issue is ported over from the upstream issue [#373](https://glotpress.trac.wordpress.org/ticket/373) by @nb.

> While working on [#362](https://glotpress.trac.wordpress.org/ticket/362), I had to decouple two main entities (originals and translations) from projects. The main value translation sets are bringing is that they couple a locale (+ variant) to a project and give us an easy ID to refer to.  
> They also allow us to customize what locales are available to which projects and under what name, which we haven't used much.  
> So, I want to explore what would a world without translation sets look like. Here is a try:  
> - It will be easier for outside people to grasp the meaning of entities in the system. Everybody knows what a project, or a locale means. I had trouble explaining what a translation sets means.
> - Translations won't have a translation_set_id, but instead locale and locale_variant (formal, informal, etc.) columns. We need this anyway if we want to decouple translations from projects (needed for #362).
> - We didn't use them for permissions anyway, we used a combination of project, locale slug and locale variant (the translation set slug).
> - It won't be hard to switch glossaries to use project + locale + variant.
> - Each project page will show all locales, which can be visually annoying on projects with small number of locales. We can make it less annoying by hiding locales without translations.
> - We won't need to create translation sets for new locale or new projects.
> - It will make querying (and partitioning) things by locale much easier.
> - We won't need to sync translation sets when branching projects.
> 
> Concerns? Holes in this idea? Any complications in sight?

**Comment by toolstack** [comment]
I think translation sets make sense for a few reasons:
- They decouple the details of the translation set from the strings/project.  So if I need to change a set's slug (or locale or whatever) I only have to update a single database row instead of every translation.
- They provide an easy and computationally cheap way of showing things like "all projects that have a locale of ar"
- The above comment about not using them for permissions, in my opinion, is a bug that needs to be fixed :wink:

Having said that there probably are some improvements we could make.  Changing "slug" to "locale variant" (or probably something even easier to understand) would be a good start.  Maybe changing their name (Translation Locale Variant?) would help.

Most of the actual functioning of sets could probably be hidden from the user while maintaining their benefits to the codebase.

**Comment by dd32** [comment]
My views on Translation Sets are rather vague, but in the end it boils down to several main points
- What is a Translation Set
- How do I query GlotPress for Projects and/or Translations and/or Originals
- How should Translation Sets actually be used

Translation sets as currently implemented link several things together `(Project, Locale, Locale Variant, Locale Name (Set Name))`, they act together to provide a `locale => project` mapping.

The problem as I see it, and at least on WordPress.org, is that we don't actually want a `locale => project` mapping. We don't want or need 1.5million translation sets (Yes, that's what we currently have) and it only makes life difficult when we want to do anything other than a basic Project/Locale listing.

As translation sets currently are:
- I can't query for specific translations in a given locale without an expensive join
- I can't determine a users contributions to a locale without an expensive join
- I can't determine all users who have contributed to a locale without a super-expensive join
- I can't perform complex queries against the projects table, as I then have to join against the translation sets table to verify that the project is translated into the locale I want, and even then, because translation sets are applied to a child (or grandchild) project (in our case) I can't easily query parent projects whose children have a specific locale
- Because we have 1.5million sets, we currently cannot cache sets in memory, as it's 1.5million copies of duplicated content (which is projected to grow to several million over the next few months)

WordPress.org has got around this in two main ways:
1. We stopped querying GlotPress tables, we now have a `project_translation_status` which has `( project_id, locale, status, translated, untranslated, etc )` fields we can query instead
2. We started storing user stats in a table that's live-updated so we can query it directly to find out what we need (This lacks per-project numbers, but we can join this off against the translations table to get what we need)
3. ..and just performing expensive queries

So to take a step back, I'd like to explain how I think Translation Sets should really be used - because I don't think they should be removed, I think they just need a re-think
1. We should assume that all projects have support for all locales, this should be the default behaviour of GlotPress and you should use a plugin if you need to block out a locale from a project
2. Translation Sets should have a schema of `( id, locale, variant, name )` there should be say 10 translation sets and 100 projects, all 100 sharing those sets (total of 10 translation sets, not 1,000).
3. Translations should have several main ID keys attached to them: `( translation_id, project_id, original_id, translation_set_id, ....)` what this means is that the translations table would gain a `project_id` field and the `translation_set_id` field would no longer be unique to `project+locale`, `Project1` and `Project2` would have `translation_set_id=1`

(A small note on 3 - Adding a project_id key would work against another proposal which suggests translations should be split away from projects, with a tables linking originals to project, and translations to sets/originals. I don't see this as being a problem, it would simply mean that the joining table would gain a field)

With the above in mind, querying for projects becomes much easier as you can assume that a locale exists for the project without multiple joins (one join per level of depth you wish to search), indexes become easier to use as the cardinality is far lower, there's only say 200 translation sets instead of 2 million, compound keys make pulling up a project+locale just as fast as currently is possible.

Now to address some things mentioned here:

> They decouple the details of the translation set from the strings/project. So if I need to change a set's slug (or locale or whatever) I only have to update a single database row instead of every translation.

In the case of WordPress.org, we have to update 10k+ translation sets if we want to change the name

> They provide an easy and computationally cheap way of showing things like "all projects that have a locale of ar"

I really question the need to do that, but if we shifted to assuming that all locales were available for all projects, then that would become moot. If a site had a need to remove/limit projects from a locale though, a plugin with it's own table would allow that query to happen, and probably more efficiently

> The above comment about not using them for permissions, in my opinion, is a bug that needs to be fixed 

Permissions are something I've not touched on yet, and is actually an important thing to note. At present permissions _should_ be specified based on a translation_set_id, however that doesn't scale as WordPress.org would need 10k+ and growing permissions, which then indicates that you have to specify it per-locale..
Permissions should, IMHO, be handled on a Project+Translation_Set basis, which under the above suggestion, is very close to how it's currently handled (It's currently `(project_id, locale) 1|en-au`, and it'd become `(project_id, set_id) 1|1` in mine, and if implemented "properly" currently it'd be `(set_id) 1500000` instead of `(project_id, set_id) 1000|9`)

**Comment by toolstack** [comment]
Reading @dd32's comment it seems to me that it's really the data model that's lacking in many of the cases.  The data is there, but it's expensive to get to.

A proper data model would take in to account many of the things above and add them to the database/code in a way that would be inexpensive to retrieve later.

An area we do have to be a little careful of is assuming wordpress.org is the only implementation of GP.  Other installs may have different requirements for things like permissions (may want to limit a user to a specific project/locale/slug like is currently possible).  I don't think that is a show stopper by any means, just something to keep in mind.

A few specific thoughts on the above...

> We should assume that all projects have support for all locales, this should be the default behaviour of GlotPress and you should use a plugin if you need to block out a locale from a project

Not a fan of this idea, for a large install like wordpress.org it probably make sense, but for smaller sites it probably doesn't.  Of course the devil would be in the implementation details so it might work for both.

It does make some things easier, but it makes others harder.  For example, when you go to a project now, you get the list of translation sets, but if you assume all locales exists for every project, you have to query the database to find out how many locale's actually have content and should be displayed (or if you maintain a list of locales per project that have entries in a table you've really just recreated the translation set system in an automated fashion).

> > They decouple the details of the translation set from the strings/project. So if I need to change a set's slug (or locale or whatever) I only have to update a single database row instead of every translation.
> 
> In the case of WordPress.org, we have to update 10k+ translation sets if we want to change the name

That assumes you want to change all 10k of them :wink:

This was mostly a comment that you didn't have to go through other tables in the database (like permissions) and update the entries in each one.  At the moment, if you change a slug for a translation set, you have to manually go and update the permissions you have associated with it.

> > The above comment about not using them for permissions, in my opinion, is a bug that needs to be fixed 
> 
> Permissions are something I've not touched on yet, and is actually an important thing to note. At present permissions should be specified based on a translation_set_id, however that doesn't scale as WordPress.org would need 10k+ and growing permissions, which then indicates that you have to specify it per-locale..

That's assuming the current permission system isn't extended.  Why not create a new permission type that defines a locale approver, then you just need one permission entry per user.

There's another discussion to be had here about the current permissions model and how to improve it which I think would handle many of these kinds of cases.

**Comment by dd32** [comment]
> An area we do have to be a little careful of is assuming wordpress.org is the only implementation of GP. Other installs may have different requirements for things like permissions (may want to limit a user to a specific project/locale/slug like is currently possible). I don't think that is a show stopper by any means, just something to keep in mind.

My comments are mostly about a Large install, it's easy to make customisations for a small install by filtering or joining, but if you want an application that can scale to thousands, you have to design it in a way that flexibility is provided for both - and the current way is geared towards small sites with no thought given for a larger install.

> That assumes you want to change all 10k of them :wink:

Yeah, if you just want to change one, then IMHO you're doing it wrong. That's not what a Translation Set should be in my mind.

**Comment by ocean90** [comment]
> For example, when you go to a project now, you get the list of translation sets, but if you assume all locales exists for every project, you have to query the database to find out how many locale's actually have content and should be displayed (or if you maintain a list of locales per project that have entries in a table you've really just recreated the translation set system in an automated fashion).

I think the assumption is to use `GP_Locales()->locales()` and to show all locales instead.

**Comment by toolstack** [comment]
That might be the assumption, but it's not a very good user interface if I have a project that only has one or two locale's with translations and I have to sift through hundreds to find them.

**Comment by dd32** [comment]
> I think the assumption is to use GP_Locales()->locales() and to show all locales instead.

I would personally keep Translation Sets around still, just not have them unique per project. As a result an install could still only have the 5 locales it's interested in. It'd be more of a `GP_Translation_Sets->get()` call as a result (or most likely `GP_Translation_Sets->get_for_project( $project )` which maps to `get()` internally with a filter for the below scenario)

> That might be the assumption, but it's not a very good user interface if I have a project that only has one or two locale's with translations and I have to sift through hundreds to find them.

In my mind, the existing UI/data is flexible, but at the expense of significant complexity as a result of it.
By simplifying it, it allows for a plugin to add the minimal complexity required - for example, hiding a locale for a project, or hiding it for all projects except X.

**Comment by ocean90** [comment]
Just stumbled upon this once again. It's nearly impossible to perform a simple query for a translation of a locale without retrieving all set IDs first.

Or to search for a original and get its translations of a locale.

> I would personally keep Translation Sets around still, just not have them unique per project

:100: :100: :100:

**Comment by toolstack** [comment]
While not truly "simple", it is just a single join, which isn't really complex either:

```
SELECT * FROM `wp_gp_translations` JOIN `wp_gp_translation_sets` ON `wp_gp_translations`.`translation_set_id` = `wp_gp_translation_sets`.id WHERE `wp_gp_translation_sets`.locale = 'ar' ;
```

**Comment by ocean90** [comment]
``` sql
EXPLAIN EXTENDED SELECT `translate_translations`.* FROM `translate_translations` JOIN `translate_translation_sets` ON `translate_translations`.`translation_set_id` = `translate_translation_sets`.id WHERE `translate_translation_sets`.locale = 'de-ch' AND `translate_translations`.`translation_0` LIKE '%foo%' LIMIT 30;
```

| id | select_type | table | type | possible_keys | key | key_len | ref | rows | filtered | Extra |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| 1 | SIMPLE | translate_translations | ALL | translation_set_id,translation_set_id_status | NULL | NULL | NULL | 32119614 | 100.00 | Using where |
| 1 | SIMPLE | translate_translation_sets | eq_ref | PRIMARY | PRIMARY | 4 | translate.translate_translations.translation_set_id | 1 | 100.00 | Using where |

**Comment by toolstack** [comment]
And you only retrieved one row from the translation sets :wink:

Of course the fact we don't index many of the columns does have an impact, we should probably review which columns have indexes. 

In your example above, all the translation rows get hit if for no other reason that the `LIKE` condition.

Even without that, since we don't index the `translation_set_id`, we'd probably still have to hit all the rows in the table to find the ones we want.
